### ISSUE TITLE
Allow sharing of chats, characters, workspaces with specific users/emails

### ISSUE THREAD

**Issue by kr3t3n** [mckaywrigley/chatbot-ui#1550]
It would be fantastic if chats, characters, prompts and workspaces can be shared with specific users via their email, for example.

**Comment by Pils10** [comment]
Or simply just via a link, like ChatGPT.

**Comment by kr3t3n** [comment]
Yeah but links can be guessed by bots. If links, would prefer if we got to choose if the links gives access to anyone with link or only the people/groups shared with

**Comment by Pils10** [comment]
Ok, get it. But it all depends on the link length and complexity of the link. For example, YouTube Video URLs are pretty short and still rarely guessed by bots (chance is about 62^11). If the current Chat ID is used for a link, like for example, "75d84ba6-26b4-4d79-a991-db763480ec85", it will be almost unguessable (maybe in a billion years, but who cares than). Nobody is going to put so much effort into brute-forcing a shared URL. Just make sure it won't get indexed by search engines and everything's safe. Implementing a shared-access auth feature would definitely take way more time than just generating a long, sharable link.

**Comment by kr3t3n** [comment]
Agree about the effort but from security perspective even the low risk is easily avoidable with simple email/username access control.

Some use cases are not possible without such.

**Comment by Pils10** [comment]
But user accounts can also be brute forced. Then you have the same problem. If you really want to get rid of this, you have to implement two-factor authentication and secure the APIs, since some of them store the queries you make and could be attacked. And still, if your hosting provider or the database-provider got hacked, all your messages are in the hands of the attacker. In conclusion, I don't think such a feature is necessary nor useful in everyday use (message-contents can still be shared with unwanted 3rd parties, even after an account system). The only reason I see for such a feature would be cooperative prompting.

**Comment by kr3t3n** [comment]
But that's on the users/orgs, not the product. It's not about complete security, but to me access control is a 'must-have' for business/enterprise.

2FA would be great too, but later.

**Comment by inzomiac** [comment]
maybe by adding a new table that will enable sharing... query chat message per user then shared table.user_id == profile.user_id. Shared table should only contain message_id, prompt_id, tools_id, etc and userID.

thats how I will implement this... in away it will be secure rather than a URL.

**Comment by xl0** [comment]
I think the way this could be done is, just to have a switch in the chat window:
![image](https://github.com/mckaywrigley/chatbot-ui/assets/439415/129f5470-8990-466e-8799-db5ff79cf029)

**Comment by aydar-musin** [comment]
really need this feature - share workspaces, assistants, models between users
